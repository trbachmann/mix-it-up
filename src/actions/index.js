export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setRecipes = (recipes) => ({
  type: 'SET_RECIPES',
  recipes
});

export const setAttribution = (attribution) => ({
  type: 'SET_ATTRIBUTION',
  attribution
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const addUserRecipe = (recipe) => ({
  type: 'ADD_USER_RECIPE',
  recipe
});

export const updateUserRecipe = (recipe) => ({
  type: 'UPDATE_USER_RECIPE',
  recipe
});

export const setStatus = (status) => ({
  type: 'SET_STATUS',
  status
});