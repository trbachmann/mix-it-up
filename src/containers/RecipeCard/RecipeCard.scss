.RecipeCard-article {
  border-radius: 8px;
  background-color: #fff;
  width: 800px;
  margin: 2em auto;
  box-shadow: 0px 1px 5px 0px rgba(0, 0, 0, 0.2),
  0px 2px 2px 0px rgba(0, 0, 0, 0.14),
  0px 3px 1px -2px rgba(0, 0, 0, 0.12);
}

.RecipeCard--p {
  font-weight: bold;
  margin: .5em 0;
}

.RecipeCard--div--flex {
  display: flex;
}

.RecipeCard--div {
  display: inline-block;
  margin: 1em 1em 1em 2em;
}

.RecipeCard--div-notes {
  @extend .RecipeCard--div;
  width: 400px;
}

.RecipeCard--anchor {
  color: $darkMaroon;
  font-size: 1.7em;
  &:hover {
    color: blue;
  }
}

.RecipeCard--img {
  border-radius: 8px;
  width: 300px;
}

.RecipeCard--button {
  background-color: $maroon;
  border-radius: 8px;
  color: #fff;
  display: block;
  font-size: 1em;
  padding: .5em 1.8em;
  &:hover {
    background-color: $darkMaroon;
  }
  margin-top: .6em;
}

.RecipeCard--textarea {
  height: 140px;
  width: 400px;
  font-size: 1em;
}

.RecipeCard--p--large {
  font-size: 1.2em;
  margin: 1.3em 0 .9em 0;
}

.RecipeCard--div--buttons {
  display: flex;
  justify-content: center;
}

.RecipeCard--icon--close {
  margin-top: .7em;
  margin-left: .5em;
  font-size: 1.2em;
  border: 0;
  &:hover {
    color: $darkMaroon;
  }
}

.RecipeCard--icon--delete {
  @extend .RecipeCard--icon--close;
  margin-left: auto;
  margin-right: .5em;
}

@media screen and (max-width: 900px) {
  .RecipeCard--div--flex {
    flex-direction: column;
  }

  .RecipeCard-article {
    width: 500px;
  }

  .RecipeCard--img {
    border-radius: 8px;
    width: 400px;
  }

  .RecipeCard--div {
    margin: auto auto 20px;
  }

  .RecipeCard--div--buttons {
    margin-bottom: 10px;
  }
}

@media screen and (max-width: 550px) {
  .RecipeCard--div--flex {
    flex-direction: column;
  }

  .RecipeCard-article {
    width: 300px;
  }

  .RecipeCard--textarea {
    width: 250px;
  }

  .RecipeCard--img {
    border-radius: 8px;
    width: 250px;
  }

  .RecipeCard--div-notes {
    width: 250px;
  }
}