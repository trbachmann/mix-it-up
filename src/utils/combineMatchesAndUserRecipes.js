export const combineMatchesAndUserRecipes = (matches, userRecipes) => {
  if (!userRecipes) {
    return matches
  }
  
  let uniqueUserRecipes = [];
  userRecipes.forEach(userRecipe => {
    let found = false;
    matches.forEach(match => {
      if (match.id === userRecipe.id) {
        found = true;
        match.notes = userRecipe.notes;
      }
    });

    if (!found) {
      uniqueUserRecipes.push(userRecipe);
    }
  });
  return [...matches, ...uniqueUserRecipes];
}