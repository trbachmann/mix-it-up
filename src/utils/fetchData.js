export const fetchData = async (url) => {
  const response = await fetch(url);
  if(!response.ok) {
    throw new Error(response.message)
  } 
   return await response.json();
}