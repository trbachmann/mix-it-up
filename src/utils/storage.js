export const setLocalStorage = (id, notes) => {
  const recipes = JSON.stringify([{ id, notes }]);
  localStorage.setItem('userRecipes', recipes);
};

export const updateLocalStorage = (id, notes) => {
  const savedRecipes = JSON.parse(localStorage.getItem('userRecipes'));
  const found = savedRecipes.find(savedRecipe => savedRecipe.id === id);
  let updatedRecipes;
  if (found) {
    updatedRecipes = savedRecipes.map(savedRecipe => {
      if (savedRecipe.id === id) {
        savedRecipe.notes = notes;
      }
      return savedRecipe
    });
  } else {
    updatedRecipes = [...savedRecipes, { id, notes }];
  }
  localStorage.setItem('userRecipes', JSON.stringify(updatedRecipes));
};

export const removeFromLocalStorage = (id) => {
  const savedRecipes = JSON.parse(localStorage.getItem('userRecipes'));
  const updatedRecipes = savedRecipes.filter(recipe => recipe.id !== id);
  localStorage.setItem('userRecipes', JSON.stringify(updatedRecipes));
};