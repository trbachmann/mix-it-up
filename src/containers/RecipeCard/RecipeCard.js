import React, { Component } from 'react';
import { connect } from 'react-redux';
import { deleteRecipeNotes, updateRecipeNotes, setStatus } from '../../actions';
import { NavLink, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import * as storage from '../../utils/storage';

export class RecipeCard extends Component {
  constructor() {
    super();
    this.state = {
      notes: ''
    }
  }

  componentDidMount = () => {
    const notes = this.props.recipe.notes;
    this.setState({ notes });
  }

  componentWillUnmount = () => {
    this.props.setStatus('');
  }

  handleChange = (event) => {
    const { value } = event.target;
    this.setState({ notes: value });
  }

  handleDelete = () => {
    const { deleteRecipeNotes, setStatus, recipe } = this.props;
    storage.removeFromLocalStorage(recipe.id);
    deleteRecipeNotes(recipe.id);
    setStatus('success');
  }

  handleSave = () => {
    const { id } = this.props.recipe;
    const { notes } = this.state;
    const { setStatus, updateRecipeNotes } = this.props;
    updateRecipeNotes(id, notes);
    if (localStorage.hasOwnProperty('userRecipes')) {
      storage.updateLocalStorage(id, notes);
    } else {
      storage.setLocalStorage(id, notes);
    }

    setStatus('success');
  }

  render() {
    const { images, name, source, totalTime } = this.props.recipe;
    const { match } = this.props;
    let style = 'RecipeCard--div--button';
    if (match.path === '/my-recipes/:id') {
      style = 'RecipeCard--div--buttons';
    } 
    return (
      <article className='RecipeCard--article'>
        <div className={style}>
          <NavLink to='/'>
            <button className='RecipeCard--icon--close'><i className='far fa-times-circle'></i></button>
          </NavLink>
          {
            (match.path === '/my-recipes/:id') && 
            <button onClick={this.handleDelete} className='RecipeCard--icon--delete'><i className="far fa-trash-alt"></i></button>
          }
        </div>
        <div className='RecipeCard--div--flex'>
          <div className='RecipeCard--div'>
            <a className='RecipeCard--anchor'
              href={source.sourceRecipeUrl}
              target='_blank'
              rel='noopener noreferrer'>
              <img className='RecipeCard--img' src={images[0].hostedLargeUrl} alt={name} />
            </a>
            <p className='RecipeCard--p'>Time: {totalTime}</p>
            <p className='RecipeCard--p'>From: {source.sourceDisplayName}</p>
          </div>
          <div className='RecipeCard--div-notes'>
            <a className='RecipeCard--anchor' 
              href={source.sourceRecipeUrl} 
              target='_blank' 
              rel='noopener noreferrer'>{name}</a>
            <p className='RecipeCard--p--large'>My Baking Notes:</p>
            <textarea 
              className='RecipeCard--textarea' 
              value={this.state.notes} 
              onChange={this.handleChange} 
              placeholder='Save your baking notes and specific oven instructrions'
            ></textarea>
            <button onClick={this.handleSave} className='RecipeCard--button'>Save</button>
          </div>

        </div>
        { (this.props.status === 'success') && <Redirect to='/'/> }
      </article>
    );
  }
}

export const mapStateToProps = (state) => ({
  status: state.status
});

export const mapDispatchToProps = (dispatch) => ({
  deleteRecipeNotes: (id) => dispatch(deleteRecipeNotes(id)),
  updateRecipeNotes: (id, notes) => dispatch(updateRecipeNotes(id, notes)),
  setStatus: (status) => dispatch(setStatus(status)),
});

export default connect(mapStateToProps, mapDispatchToProps)(RecipeCard);

RecipeCard.propTypes = {
  recipe: PropTypes.object,
  match: PropTypes.object,
  status: PropTypes.string,
  deleteRecipeNotes: PropTypes.func,
  updateRecipeNotes: PropTypes.func,
  setStatus: PropTypes.func,
}