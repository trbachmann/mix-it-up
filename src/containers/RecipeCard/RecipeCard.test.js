import React from 'react';
import { shallow } from 'enzyme';
import { RecipeCard, mapStateToProps, mapDispatchToProps } from './RecipeCard';
import * as data from '../../mockData';
import { deleteRecipeNotes, updateRecipeNotes, setStatus } from '../../actions';
import * as storage from '../../utils/storage';

describe('RecipeCard', () => {

  describe('RecipeCard component', () => {
    storage.setLocalStorage = jest.fn();
    storage.updateLocalStorage = jest.fn();
    storage.removeFromLocalStorage = jest.fn();
    const mockProps = {
      recipe: data.mockRecipe,
      match: { path: '/' },
      status: '',
      deleteRecipeNotes: jest.fn(),
      updateRecipeNotes: jest.fn(),
      setStatus: jest.fn(),
    };

    const mockPropsWithRecipeNotes = {
      recipe: data.mockRecipeWithUserNote,
      match: { path: '/my-recipes/:id' },
      status: '',
      deleteRecipeNotes: jest.fn(),
      updateRecipeNotes: jest.fn(),
      setStatus: jest.fn(),
    }

    let wrapper;
    let wrapperMyRecipes;

    beforeEach(() => {
      wrapper = shallow(<RecipeCard {...mockProps}/>);
      wrapperMyRecipes = shallow(<RecipeCard {...mockPropsWithRecipeNotes}/>);
    });

    afterEach(() => {
      localStorage.removeItem('userRecipes');
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot', () => {
      

      expect(wrapperMyRecipes).toMatchSnapshot();
    });

    describe('componentDidMount', () => {
      it('should set state with notes', () => {
        const expected = data.mockRecipeWithUserNote.notes;
        wrapperMyRecipes.instance().componentDidMount();
        expect(wrapperMyRecipes.state('notes')).toEqual(expected);
      });
    });

    describe('componentWillUnmount', () => {
      it('should call setStatus', () => {
        wrapper.instance().componentWillUnmount();
        expect(mockProps.setStatus).toHaveBeenCalledWith('');
      });
    });

    describe('handleChange', () => {
      it('should set state with notes when a user updates notes', () => {
        expect(wrapper.state('notes')).toEqual('');
        const mockEvent = {
          target: { value: 'Add an extra 1/2 cup of peanut butter' }
        }
        wrapper.find('textarea').simulate('change', mockEvent);
        const expected = 'Add an extra 1/2 cup of peanut butter';
        expect(wrapper.state('notes')).toEqual(expected);
      });
    });

    describe('handleSave', () => {
      it('should call updateRecipeNotes with an id and notes', () => {
        const expectedId = mockProps.recipe.id;
        const expectedNotes = 'Add an extra 1/2 cup of peanut butter';
        wrapper.setState({ notes:  expectedNotes });
        wrapper.find('.RecipeCard--button').simulate('click');
        expect(mockProps.updateRecipeNotes).toHaveBeenCalledWith(expectedId, expectedNotes);
      });

      it('should call setLocalStorage if nothing saved', () => {
        const expectedId = mockProps.recipe.id;
        const expectedNotes = 'Add an extra 1/2 cup of peanut butter';
        wrapper.setState({ notes: expectedNotes });
        wrapper.find('.RecipeCard--button').simulate('click');
        expect(storage.setLocalStorage).toHaveBeenCalledWith(expectedId, expectedNotes);
      });

      it('should call updateLocalStorage is there are already recipes saved', () => {
        localStorage.setItem('userRecipes', JSON.stringify(data.mockUserRecipeFromStorage));
        const expectedId = mockProps.recipe.id;
        const expectedNotes = 'Add an extra 1/2 cup of peanut butter';
        wrapper.setState({ notes: expectedNotes });
        wrapper.find('.RecipeCard--button').simulate('click');
        expect(storage.updateLocalStorage).toHaveBeenCalledWith(expectedId, expectedNotes);
      });


      it('should call setStatus with success', () => {
        const expectedNotes = 'Add an extra 1/2 cup of peanut butter';
        wrapper.setState({ notes: expectedNotes });
        wrapper.find('.RecipeCard--button').simulate('click');
        expect(mockProps.setStatus).toHaveBeenCalledWith('success');
      });
    });

    describe('handleDelete', () => {

      it('should call removeFromLocalStorage', () => {
        wrapper.instance().handleDelete()
        const expected = mockProps.recipe.id;        
        expect(storage.removeFromLocalStorage).toHaveBeenCalledWith(expected);
      });

      it('should call deleteRecipeNotes with the recipe id', () => {
        wrapper.instance().handleDelete()
        const expected = mockProps.recipe.id;
        expect(mockProps.deleteRecipeNotes).toHaveBeenCalledWith(expected);
      });

      it('should set status with success', () => {
        wrapper.instance().handleDelete();
        expect(mockProps.setStatus).toHaveBeenCalledWith('success');
      });
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with status and desserts', () => {
      const mockState = {
        desserts: data.mockRecipes,
        status: '',
        error: '',
        isLoading: false
      };
      const expected = {
        status: ''
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();

    it('should call dispatch with a deleteRecipeNotes action', () => {
      const actionToDispatch = deleteRecipeNotes('Oatmeal-Rasin-Cookies-34576');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.deleteRecipeNotes('Oatmeal-Rasin-Cookies-34576');
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with an updateRecipeNotes action', () => {
      const mockId = 'Oatmeal-Rasin-Cookies-34576';
      const mockNotes = 'Increase bakimg time to 15mins'
      const actionToDispatch = updateRecipeNotes(mockId, mockNotes);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.updateRecipeNotes(mockId, mockNotes);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should call dispatch with a setStatus action', () => {
      const actionToDispatch = setStatus('success');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setStatus('success');
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});