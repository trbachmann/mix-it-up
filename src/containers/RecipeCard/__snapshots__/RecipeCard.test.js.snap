// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard RecipeCard component should match the snapshot 1`] = `
<article
  className="RecipeCard--article"
>
  <div
    className="RecipeCard--div--button"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      to="/"
    >
      <button
        className="RecipeCard--icon--close"
      >
        <i
          className="far fa-times-circle"
        />
      </button>
    </NavLink>
  </div>
  <div
    className="RecipeCard--div--flex"
  >
    <div
      className="RecipeCard--div"
    >
      <a
        className="RecipeCard--anchor"
        href="https://juliasalbum.com/easy-peanut-butter-cookies/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="The Best (and Easy) Peanut Butter Cookies"
          className="RecipeCard--img"
          src="https://lh3.googleusercontent.com/IL4SAQ143siXT-QfTZjeNORYdW5BTxDXLdQw-Ix1-rBMLlTfHKhZZFajd3myiwdIxndI9Kv3mpdzetFJPz7qfQ=s360"
        />
      </a>
      <p
        className="RecipeCard--p"
      >
        Time: 
        40 min
      </p>
      <p
        className="RecipeCard--p"
      >
        From: 
        Julia's Album
      </p>
    </div>
    <div
      className="RecipeCard--div-notes"
    >
      <a
        className="RecipeCard--anchor"
        href="https://juliasalbum.com/easy-peanut-butter-cookies/"
        rel="noopener noreferrer"
        target="_blank"
      >
        The Best (and Easy) Peanut Butter Cookies
      </a>
      <p
        className="RecipeCard--p--large"
      >
        My Baking Notes:
      </p>
      <textarea
        className="RecipeCard--textarea"
        onChange={[Function]}
        placeholder="Save your baking notes and specific oven instructrions"
        value=""
      />
      <button
        className="RecipeCard--button"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
</article>
`;

exports[`RecipeCard RecipeCard component should match the snapshot 2`] = `
<article
  className="RecipeCard--article"
>
  <div
    className="RecipeCard--div--buttons"
  >
    <NavLink
      activeClassName="active"
      aria-current="page"
      to="/"
    >
      <button
        className="RecipeCard--icon--close"
      >
        <i
          className="far fa-times-circle"
        />
      </button>
    </NavLink>
    <button
      className="RecipeCard--icon--delete"
      onClick={[Function]}
    >
      <i
        className="far fa-trash-alt"
      />
    </button>
  </div>
  <div
    className="RecipeCard--div--flex"
  >
    <div
      className="RecipeCard--div"
    >
      <a
        className="RecipeCard--anchor"
        href="https://juliasalbum.com/easy-peanut-butter-cookies/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <img
          alt="The Best (and Easy) Peanut Butter Cookies"
          className="RecipeCard--img"
          src="https://lh3.googleusercontent.com/IL4SAQ143siXT-QfTZjeNORYdW5BTxDXLdQw-Ix1-rBMLlTfHKhZZFajd3myiwdIxndI9Kv3mpdzetFJPz7qfQ=s360"
        />
      </a>
      <p
        className="RecipeCard--p"
      >
        Time: 
        40 min
      </p>
      <p
        className="RecipeCard--p"
      >
        From: 
        Julia's Album
      </p>
    </div>
    <div
      className="RecipeCard--div-notes"
    >
      <a
        className="RecipeCard--anchor"
        href="https://juliasalbum.com/easy-peanut-butter-cookies/"
        rel="noopener noreferrer"
        target="_blank"
      >
        The Best (and Easy) Peanut Butter Cookies
      </a>
      <p
        className="RecipeCard--p--large"
      >
        My Baking Notes:
      </p>
      <textarea
        className="RecipeCard--textarea"
        onChange={[Function]}
        placeholder="Save your baking notes and specific oven instructrions"
        value="Increase peanut butter by 1/2 cup and increase vanilla 2 teaspoons"
      />
      <button
        className="RecipeCard--button"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
</article>
`;
